shell/renderer/api/atom_api_context_bridge.cc

```diff
@@ -0,0 +1,408 @@
+    return CreateProxyForAPI(
+               mate::Dictionary(source->GetIsolate(),
+                                v8::Local<v8::Object>::Cast(value)),
+               source, destination, store)
+        .GetHandle();
+  }
+
+  // Serializable objects
+  // TODO(MarshallOfSound): Use the V8 serializer so we can remove the special
+  // null / undefiend handling
+  if (value->IsUndefined()) {
+    v8::Context::Scope destination_context_scope(destination);
+    return v8::Undefined(destination->GetIsolate());
+  }
+
+  if (value->IsNull()) {
+    v8::Context::Scope destination_context_scope(destination);
+    return v8::Null(destination->GetIsolate());
+  }
+
+  base::Value ret;
+  {
+    v8::Context::Scope source_context_scope(source);
+    // TODO(MarshallOfSound): What do we do if serialization fails? Throw an
+    // error here?
+    if (!mate::ConvertFromV8(source->GetIsolate(), value, &ret))
+      return v8::Null(destination->GetIsolate());
+  }
+
+  v8::Context::Scope destination_context_scope(destination);
+  return mate::ConvertToV8(destination->GetIsolate(), ret);
+}
+
+v8::Local<v8::Value> ProxyFunctionWrapper(RenderFramePersistenceStore* store,
+                                          size_t func_id,
+                                          mate::Arguments* args) {
+  // Context the proxy function was called from
+  v8::Local<v8::Context> calling_context = args->isolate()->GetCurrentContext();
+  // Context the function was created in
+  v8::Local<v8::Context> func_owning_context =
+      std::get<1>(store->functions()[func_id]).Get(args->isolate());
+
+  v8::Context::Scope func_owning_context_scope(func_owning_context);
+  v8::Local<v8::Function> func =
+      (std::get<0>(store->functions()[func_id])).Get(args->isolate());
+
+  std::vector<v8::Local<v8::Value>> original_args;
+  std::vector<v8::Local<v8::Value>> proxied_args;
+  args->GetRemaining(&original_args);
+  for (auto value : original_args) {
+    proxied_args.push_back(PassValueToOtherContext(
+        calling_context, func_owning_context, value, store));
+  }
+
+  v8::MaybeLocal<v8::Value> maybe_return_value;
+  bool did_error = false;
+  std::string error_message;
+  {
+    v8::TryCatch try_catch(args->isolate());
+    maybe_return_value = func->Call(func_owning_context, func,
+                                    proxied_args.size(), proxied_args.data());
+    if (try_catch.HasCaught()) {
+      did_error = true;
+      auto message = try_catch.Message();
+
+      if (message.IsEmpty() ||
+          !mate::ConvertFromV8(args->isolate(), message->Get(),
+                               &error_message)) {
+        error_message =
+            "An unknown exception occurred in the isolated context, an error "
+            "occurred but a valid exception was not thrown.";
+      }
+    }
+  }
+
+  if (did_error) {
+    v8::Context::Scope calling_context_scope(calling_context);
+    args->ThrowError(error_message);
```

Comment:
Should we [sanitize errors](https://cs.chromium.org/chromium/src/third_party/blink/renderer/bindings/core/v8/sanitize_script_errors.h?q=sanitizescripterror&sq=package:chromium&g=0&l=10)?

Reply:
These errors don't flow across origins right? Technically it's the same origin but a different context

Reply:
The reason for the errors not to flow across origins is because you can inadvertently leak information through an error, and information isn't supposed to flow across origins unintentionally. I think the same is true of errors from the isolated world, right? I'd hesitate to say that the preload content has the same origin as the page, given the code comes from somewhere else and is controlled by something else. (I think in Chrome the isolated context is treated as having a `chrome-extension://` origin..? Maybe?)

I suppose it's not as much of a concern as cross-origin stuff, because the error was generated in the same process as the page, so if the page really wanted access to that information, [it could have it](https://spectreattack.com).

Reply:
I don't think error santization is what we want, partly because of the whole "they're kind of the same origin" and also because of this:

> But in the promise unhandled rejection case error existence is hidden when kSanitize is specified.

I.e. promise rejections are not supposed to be transferred when error sanitization is used.  That isn't that helpful for a bridge like this.

I think the current measure of stripping the stack trace and only transferring the message hides enough information for now.